File: apps/dashboard/src-tauri/src/commands/macos_shell.rs
//! Lightweight macOS shell bridges used by full-editor mode.
//!
//! These commands are NOT side-panel specific — they're called from the
//! workspace toolbar (open in editor, reveal in Finder), the settings page
//! (folder picker, install CLI), and the dashboard adapter. They were
//! historically defined in `commands/ide.rs` next to the side-panel
//! window-focusing code; that file was deleted when side-panel mode was
//! split into its own repo (`band-app/sidepanel`), so the bits we still
//! need live here.

/// Opens a macOS folder picker via `osascript` and returns the chosen
/// POSIX path, or `Ok(None)` if the user cancelled.
#[tauri::command]
pub fn pick_folder() -> Result<Option<String>, String> {
    let output = std::process::Command::new("osascript")
        .args([
            "-e",
            r#"set theFolder to choose folder with prompt "Select a git repository"
return POSIX path of theFolder"#,
        ])
        .output()
        .map_err(|e| format!("Failed to open folder picker: {e}"))?;

    if output.status.success() {
        let path = String::from_utf8_lossy(&output.stdout).trim().to_string();
        if path.is_empty() {
            Ok(None)
        } else {
            Ok(Some(path))
        }
    } else {
        Ok(None)
    }
}

/// Reveals a path in Finder via `open`.
#[tauri::command]
pub fn reveal_in_finder(path: String) -> Result<(), String> {
    std::process::Command::new("open")
        .arg(&path)
        .output()
        .map_err(|e| format!("Failed to open Finder: {e}"))?;
    Ok(())
}

/// Checks whether a macOS application is installed by looking in common
/// locations (/Applications, /System/Applications, ~/Applications) and
/// falling back to `which` for CLI tools.
#[tauri::command]
pub fn check_app_exists(app_name: String) -> bool {
    let mut locations = vec![
        format!("/Applications/{app_name}.app"),
        format!("/System/Applications/{app_name}.app"),
    ];

    if let Ok(home) = std::env::var("HOME") {
        locations.push(format!("{home}/Applications/{app_name}.app"));
    }

    for location in &locations {
        if std::path::Path::new(location).exists() {
            return true;
        }
    }

    // Fallback: check if a CLI binary exists in PATH
    std::process::Command::new("which")
        .arg(&app_name)
        .output()
        .is_ok_and(|output| output.status.success())
}

/// Opens a path with a specific macOS application.
#[tauri::command]
pub fn open_with_app(path: String, app_name: String) -> Result<(), String> {
    std::process::Command::new("open")
        .arg("-a")
        .arg(&app_name)
        .arg(&path)
        .output()
        .map_err(|e| format!("Failed to open with {app_name}: {e}"))?;
    Ok(())
}

/// Install the CLI by creating a symlink with administrator privileges.
/// Runs `osascript` to show a macOS admin password dialog — this works
/// because the Tauri app is the foreground GUI process (unlike the web
/// server, which can't reliably show GUI dialogs).
#[tauri::command]
pub fn install_cli(binary_path: String, symlink_path: String) -> Result<(), String> {
    let cmd = format!(
        "ln -sf '{}' '{}'",
        binary_path.replace('\'', "'\\''"),
        symlink_path.replace('\'', "'\\''")
    );
    let script = format!(
        "do shell script \"{}\" with administrator privileges",
        cmd.replace('\\', "\\\\").replace('"', "\\\"")
    );
    let output = std::process::Command::new("osascript")
        .args(["-e", &script])
        .output()
        .map_err(|e| format!("Failed to run osascript: {e}"))?;

    if output.status.success() {
        Ok(())
    } else {
        let stderr = String::from_utf8_lossy(&output.stderr);
        if stderr.contains("User canceled") || stderr.contains("-128") {
            Err("Admin password prompt cancelled".to_string())
        } else {
            Err(format!("Failed to install CLI: {stderr}"))
        }
    }
}

File: apps/desktop/src/main/ipc/macos-shell.ts
/**
 * macOS shell bridges.
 *
 *   - `pickFolder` — system folder picker (cross-platform via Electron's `dialog`).
 *   - `pickFile` — system file picker for opening external files (cross-platform).
 *   - `pickSaveFile` — system "Save As" picker for writing a new file to disk.
 *   - `revealInFinder` — open the path in the platform's file manager.
 *   - `checkAppExists` — look in /Applications and friends, fall back to `which`.
 *   - `openWithApp` — `open -a <app> <path>` (macOS only).
 *   - `installCli` — symlink the CLI binary via `osascript` admin prompt (macOS only).
 *   - `openExternal` — `shell.openExternal(url)` (cross-platform).
 */

import { spawn, spawnSync } from "node:child_process";
import { existsSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
import { type BrowserWindow, dialog, shell } from "electron";

import { type CliPathOptions, resolveCliBinary } from "../services/cli-paths.js";
import { resolveSaveDialogSeed, writeSavedFile } from "./save-helpers.js";

const NOT_SUPPORTED = "Not supported on this platform";

// ---------------------------------------------------------------------------
// pickFolder
// ---------------------------------------------------------------------------

/**
 * Open the system folder picker. Returns the absolute POSIX path or `null`
 * when the user cancels. Anchored to `parent` so the dialog is sheet-style
 * on macOS and modal-relative on other platforms.
 */
export async function pickFolder(parent: BrowserWindow | null): Promise<string | null> {
  const opts = {
    title: "Select a git repository",
    properties: ["openDirectory"] as Array<"openDirectory">,
  };
  const result = parent
    ? await dialog.showOpenDialog(parent, opts)
    : await dialog.showOpenDialog(opts);
  if (result.canceled || result.filePaths.length === 0) return null;
  return result.filePaths[0] ?? null;
}

// ---------------------------------------------------------------------------
// pickFile
// ---------------------------------------------------------------------------

/**
 * Open the system file picker for opening a single file from anywhere on the
 * local filesystem. Returns the absolute POSIX path or `null` when the user
 * cancels. Anchored to `parent` so the dialog is sheet-style on macOS and
 * modal-relative on other platforms.
 *
 * Backs the editor's "Open File…" action — lets a user open files
 * that sit outside the current workspace root.
 */
export async function pickFile(parent: BrowserWindow | null): Promise<string | null> {
  const opts = {
    title: "Open File",
    properties: ["openFile"] as Array<"openFile">,
  };
  const result = parent
    ? await dialog.showOpenDialog(parent, opts)
    : await dialog.showOpenDialog(opts);
  if (result.canceled || result.filePaths.length === 0) return null;
  return result.filePaths[0] ?? null;
}

// ---------------------------------------------------------------------------
// pickSaveFile
// ---------------------------------------------------------------------------

/**
 * Defensive ceiling on the buffer the renderer can hand off in a
 * single save call. The Electron model is trusted (the renderer is
 * Band's own code), so this isn't a security boundary — it's a UX
 * safety net for accidental pastes (multi-MB log files, an image
 * dropped in as base64, a minified bundle) that would otherwise
 * sit in IPC shared memory, the main-process heap, and the temp
 * file concurrently while the disk write stalls the event loop.
 *
 * 100 MB matches the rough ceiling above which the OS save dialog
 * itself becomes the bottleneck rather than the IPC. The renderer
 * receives the rejection cleanly through the existing error path.
 */
const SAVE_CONTENT_MAX_BYTES = 100 * 1024 * 1024;

/**
 * Open the system "Save As" picker and write the supplied `content` to the
 * chosen path. Returns the absolute path of the saved file, or `null` when
 * the user cancels.
 *
 * Backs the editor's "Save untitled tab" flow: an untitled buffer lives
 * entirely in the renderer until the user picks a destination — this
 * bridge surfaces the native save dialog and persists the buffer in a
 * single IPC round-trip (one user-visible Save operation). The renderer
 * then transitions the tab from "untitled" to file-backed using the
 * returned path. The write itself goes through `writeSavedFile`, which
 * uses a write-to-temp + rename pattern to avoid mid-write truncation —
 * see that function for the atomicity contract.
 *
 * Anchored to `parent` so the dialog is sheet-style on macOS and modal-
 * relative on other platforms. `defaultPath` and `defaultName` seed the
 * dialog's starting location and filename (e.g. "Untitled-1.txt") so the
 * user only has to type when overriding.
 */
export async function pickSaveFile(
  parent: BrowserWindow | null,
  args: { content: string; defaultName?: string; defaultPath?: string },
): Promise<string | null> {
  // Defensive size check — see `SAVE_CONTENT_MAX_BYTES`. Using
  // `Buffer.byteLength` (not `string.length`) so the cap reflects
  // the on-disk UTF-8 byte count, not the JS string code-unit count.
  if (Buffer.byteLength(args.content, "utf8") > SAVE_CONTENT_MAX_BYTES) {
    throw new Error(
      `File is too large to save via this flow (over ${SAVE_CONTENT_MAX_BYTES / (1024 * 1024)} MB). ` +
        "Split the content into smaller files or save through a terminal.",
    );
  }

  const seed = resolveSaveDialogSeed(args);
  const opts = {
    title: "Save As",
    defaultPath: seed,
    properties: ["createDirectory", "showOverwriteConfirmation"] as Array<
      "createDirectory" | "showOverwriteConfirmation"
    >,
  };
  const result = parent
    ? await dialog.showSaveDialog(parent, opts)
    : await dialog.showSaveDialog(opts);
  if (result.canceled || !result.filePath) return null;

  // `dialog.showSaveDialog` with `showOverwriteConfirmation` already
  // handled the overwrite prompt, so we trust the path and persist
  // the bytes. Errors propagate up through the IPC chain. `await` is
  // load-bearing — `writeSavedFile` is async to keep the Electron main
  // process event loop responsive during the disk write.
  await writeSavedFile(result.filePath, args.content);
  return result.filePath;
}

// ---------------------------------------------------------------------------
// revealInFinder
// ---------------------------------------------------------------------------

export async function revealInFinder(path: string): Promise<void> {
  // `shell.openPath` opens the folder/file in the system file manager and
  // returns an error string on failure (empty string on success).
  const err = await shell.openPath(path);
  if (err && err.length > 0) {
    throw new Error(`Failed to open Finder: ${err}`);
  }
}

// ---------------------------------------------------------------------------
// checkAppExists
// ---------------------------------------------------------------------------

/**
 * Check if a macOS app bundle (or CLI binary) is installed. Returns false on
 * non-macOS for `.app` lookups and falls back to `which` on Unix-like systems.
 *
 * Mirrors the Rust impl: tries /Applications, /System/Applications, ~/Applications,
 * then `which <name>`.
 */
export function checkAppExists(appName: string): boolean {
  const home = homedir();
  const candidates = [
    `/Applications/${appName}.app`,
    `/System/Applications/${appName}.app`,
    join(home, "Applications", `${appName}.app`),
  ];
  for (const path of candidates) {
    if (existsSync(path)) return true;
  }
  // Fallback: `which` (Unix) / `where` (Windows) for CLI tools.
  const cmd = process.platform === "win32" ? "where" : "which";
  try {
    const result = spawnSync(cmd, [appName], {
      encoding: "utf8",
      stdio: ["ignore", "pipe", "ignore"],
    });
    return result.status === 0;
  } catch {
    return false;
  }
}

// ---------------------------------------------------------------------------
// openWithApp
// ---------------------------------------------------------------------------

export async function openWithApp(path: string, appName: string): Promise<void> {
  if (process.platform !== "darwin") {
    throw new Error(NOT_SUPPORTED);
  }
  await new Promise<void>((resolve, reject) => {
    const proc = spawn("open", ["-a", appName, path], { stdio: "ignore" });
    proc.on("error", (err) => reject(new Error(`Failed to open with ${appName}: ${err.message}`)));
    proc.on("exit", (code) => {
      if (code === 0) resolve();
      else reject(new Error(`open -a ${appName} failed with exit code ${code}`));
    });
  });
}

// ---------------------------------------------------------------------------
// installCli
// ---------------------------------------------------------------------------

/**
 * Create a symlink with administrator privileges. Pops a macOS password
 * dialog via osascript. Direct port of the Rust impl, including the same
 * single-quote escaping (replace `'` with `'\''` then JSON-quote).
 *
 * `paths` carries the host context so the handler can resolve the bundled
 * sidecar (issue #364): in packaged builds the binary lives at
 * `process.resourcesPath/binaries/band` regardless of what the renderer
 * supplies, so we trust the resolver over the renderer's argument. In dev,
 * we fall through to the renderer-supplied path because the web server's
 * own `findCliBinary` already knows how to locate the cargo target.
 *
 * Resolving in the main process keeps the symlink target inside the trust
 * boundary of the desktop shell — the renderer can't trick us into linking
 * `/usr/local/bin/band` to an attacker-controlled path.
 */
export async function installCli(
  binaryPath: string,
  symlinkPath: string,
  paths?: CliPathOptions,
): Promise<void> {
  if (process.platform !== "darwin") {
    throw new Error(NOT_SUPPORTED);
  }
  // In packaged mode the bundled sidecar is the only trustworthy source.
  // In dev we accept the renderer-supplied path (web server already
  // resolved a cargo target) but still prefer the resolver if it finds a
  // local cargo build — keeps behaviour consistent across `pnpm dev:desktop`
  // and a packaged `Band.app`.
  const resolved = paths ? resolveCliBinary(paths) : null;
  const sourcePath = paths?.isPackaged
    ? (resolved ?? throwMissingBundledBinary())
    : (resolved ?? binaryPath);

  const escapedBinary = sourcePath.replaceAll("'", "'\\''");
  const escapedSymlink = symlinkPath.replaceAll("'", "'\\''");
  const cmd = `ln -sf '${escapedBinary}' '${escapedSymlink}'`;
  // The osascript `do shell script` string is double-quote delimited, so we
  // escape backslashes then double-quotes.
  const script = `do shell script "${cmd
    .replaceAll("\\", "\\\\")
    .replaceAll('"', '\\"')}" with administrator privileges`;

  await new Promise<void>((resolve, reject) => {
    const proc = spawn("osascript", ["-e", script], { stdio: ["ignore", "pipe", "pipe"] });
    let stderr = "";
    proc.stderr?.on("data", (chunk: Buffer) => {
      stderr += chunk.toString();
    });
    proc.on("error", (err) => reject(new Error(`Failed to run osascript: ${err.message}`)));
    proc.on("exit", (code) => {
      if (code === 0) {
        resolve();
        return;
      }
      if (stderr.includes("User canceled") || stderr.includes("-128")) {
        reject(new Error("Admin password prompt cancelled"));
        return;
      }
      reject(new Error(`Failed to install CLI: ${stderr || "unknown error"}`));
    });
  });
}

/**
 * Bail out when a packaged build is missing the bundled CLI binary. This is
 * a deployment error (`build:cli:desktop` was skipped before
 * `electron-builder`) — there is no useful fallback in a sandboxed `.app`.
 */
function throwMissingBundledBinary(): never {
  throw new Error(
    "Bundled CLI binary not found in the packaged app. " +
      "Run `pnpm build:cli:desktop` before electron-builder.",
  );
}

// ---------------------------------------------------------------------------
// openExternal
// ---------------------------------------------------------------------------

/**
 * Open a URL in the user's default browser. Backs the `open_external` IPC
 * channel called from `apps/web/src/lib/open-external-url.ts`.
 */
export async function openExternal(url: string): Promise<void> {
  await shell.openExternal(url);
}
